internal/command/jsonformat/change/primitive.go

```diff
@@ -3,46 +3,60 @@ package change
 import (
 	"fmt"
 
+	"github.com/zclconf/go-cty/cty"
+
 	"github.com/hashicorp/terraform/internal/plans"
 )
 
-func Primitive(before, after *string) Renderer {
+func Primitive(before, after interface{}, ctype cty.Type) Renderer {
 	return &primitiveRenderer{
 		before: before,
 		after:  after,
+		ctype:  ctype,
 	}
 }
 
 type primitiveRenderer struct {
 	NoWarningsRenderer
 
-	before *string
-	after  *string
+	before interface{}
+	after  interface{}
+	ctype  cty.Type
 }
 
-func (renderer primitiveRenderer) Render(result Change, indent int, opts RenderOpts) string {
-	var beforeValue, afterValue string
+func (renderer primitiveRenderer) Render(change Change, indent int, opts RenderOpts) string {
+	beforeValue := renderPrimitiveValue(renderer.before, renderer.ctype)
+	afterValue := renderPrimitiveValue(renderer.after, renderer.ctype)
 
-	if renderer.before != nil {
-		beforeValue = *renderer.before
-	} else {
-		beforeValue = "[dark_gray]null[reset]"
+	switch change.action {
+	case plans.Create:
+		return fmt.Sprintf("%s%s", afterValue, change.forcesReplacement())
+	case plans.Delete:
+		return fmt.Sprintf("%s%s%s", beforeValue, change.nullSuffix(opts.overrideNullSuffix), change.forcesReplacement())
+	case plans.NoOp:
+		return fmt.Sprintf("%s%s", beforeValue, change.forcesReplacement())
+	default:
+		return fmt.Sprintf("%s [yellow]->[reset] %s%s", beforeValue, afterValue, change.forcesReplacement())
 	}
+}
 
-	if renderer.after != nil {
-		afterValue = *renderer.after
-	} else {
-		afterValue = "[dark_gray]null[reset]"
+func renderPrimitiveValue(value interface{}, t cty.Type) string {
+	switch value.(type) {
+	case nil:
+		return "[dark_gray]null[reset]"
 	}
 
-	switch result.action {
-	case plans.Create:
-		return fmt.Sprintf("%s%s", afterValue, result.forcesReplacement())
-	case plans.Delete:
-		return fmt.Sprintf("%s%s%s", beforeValue, result.nullSuffix(opts.overrideNullSuffix), result.forcesReplacement())
-	case plans.NoOp:
-		return fmt.Sprintf("%s%s", beforeValue, result.forcesReplacement())
+	switch {
+	case t == cty.String:
+		return fmt.Sprintf("\"%s\"", value.(string))
```

Comment:
Unless I'm misunderstanding this, I don't think this copes correctly with strings which contain characters that need to be escaped (e.g. `"`).

We had previously fudged this with the `%q` specifier, but I think @apparentlymart recently added an HCL quote-string function. I can't find it right now unfortunately.

Reply:
I've updated this with `%q`, longer term I've made a note about the function buried in HCL: https://github.com/hashicorp/terraform/blob/main/internal/addrs/instance_key.go#L147

And we can make that publicly available somehow and update this, and the escaped keys for objects and blocks.